File: 07/Java/TripleExpression/CheckedDivide.java
package TripleExpression;

public class CheckedDivide extends BinaryOperator implements TripleExpression {
    public CheckedDivide(TripleExpression first, TripleExpression second) {
        super(first, second);
    }

    protected void check(int a, int b) throws Exception {
        if(a == Integer.MIN_VALUE && b == -1) {
            throw new OverflowException();
        }
        if (b == 0) {
            throw new ZeroDivisionException();
        }
    }
    protected int apply(int a, int b) throws Exception {
        check(a, b);
        return a / b;
    }
}

File: 07/CSharp/TripleExpressionApp/TripleExpressionApp/CheckedDivide.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace TripleExpressionApp
{
    public class CheckedDivide : BinaryOperator
    {
        public CheckedDivide(ITripleExpression first, ITripleExpression second)
        {
            fOp = first;
            sOp = second;
        }

        protected void Check(double a, double b)
        {
            if (a == double.MinValue && b == -1)
            {
                throw new OverflowException();
            }
            if (b == 0)
            {
                throw new ZeroDivisionException();
            }
        }

        override protected double Apply(double a, double b)
        {
            Check(a, b);
            return a / b;
        }
    }
}
